File: 8vim/src/main/java/inc/flide/vim8/structures/KeyboardData.java
package inc.flide.vim8.structures;

import inc.flide.vim8.structures.yaml.Layout;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class KeyboardData {
    private final Map<List<Integer>, KeyboardAction> actionMap = new HashMap<>();
    private final CharacterSet[] characterSets = new CharacterSet[Constants.MAX_LAYERS + 1];
    public int layoutPositions = 0;
    public int sectors = 1;
    private int totalLayers = 0;
    private Layout.LayoutInfo info;

    public Map<List<Integer>, KeyboardAction> getActionMap() {
        return actionMap;
    }

    public void addActionMap(List<Integer> movementSequence, KeyboardAction keyboardAction) {
        this.actionMap.put(movementSequence, keyboardAction);
    }

    public void addAllToActionMap(Map<List<Integer>, KeyboardAction> actionMapAddition) {
        this.actionMap.putAll(actionMapAddition);
    }

    public int getTotalLayers() {
        return totalLayers;
    }

    public String getLowerCaseCharacters(int layer) {
        if (layer < 0 || layer > totalLayers || characterSets[layer] == null) {
            return "";
        }
        return characterSets[layer].getLowerCaseCharacters();
    }


    public void setLowerCaseCharacters(String lowerCaseCharacters, int layer) {
        if (layer < 0 || layer > Constants.MAX_LAYERS) {
            return;
        }

        updateCharacterSets(layer);
        characterSets[layer].setLowerCaseCharacters(lowerCaseCharacters);
    }

    public String getUpperCaseCharacters(int layer) {
        if (layer < 0 || layer > totalLayers || characterSets[layer] == null) {
            return "";
        }
        return characterSets[layer].getUpperCaseCharacters();
    }


    public void setUpperCaseCharacters(String upperCaseCharacters, int layer) {
        if (layer < 0 || layer > Constants.MAX_LAYERS) {
            return;
        }

        updateCharacterSets(layer);
        characterSets[layer].setUpperCaseCharacters(upperCaseCharacters);
    }

    public int findLayer(List<Integer> movementSequence) {
        KeyboardAction action = actionMap.get(movementSequence);
        if (action == null) {
            return Constants.DEFAULT_LAYER;
        }
        return action.getLayer();
    }

    private void updateCharacterSets(int layer) {
        totalLayers = Math.max(totalLayers, layer);
        if (characterSets[layer] == null) {
            characterSets[layer] = new CharacterSet();
        }
    }

    public Layout.LayoutInfo getInfo() {
        return info;
    }

    public void setInfo(Layout.LayoutInfo info) {
        this.info = info;
    }
}

File: 8vim/src/main/kotlin/inc/flide/vim8/models/KeyboardData.kt
package inc.flide.vim8.models

import arrow.core.Option
import arrow.core.elementAtOrNone
import arrow.core.fold
import arrow.core.none
import arrow.core.some
import arrow.optics.dsl.index
import arrow.optics.optics
import arrow.optics.typeclasses.Index
import inc.flide.vim8.models.yaml.Layer
import inc.flide.vim8.models.yaml.LayoutInfo

const val CHARACTER_SET_SIZE = 4 * 4 * 2 // 4 sectors, 2 parts, 4 characters per parts

@optics
data class KeyboardData(
    val actionMap: Map<MovementSequence, KeyboardAction> = HashMap(),
    val characterSets: List<CharacterSet> = List(LayerLevel.visibleLayers.size) { CharacterSet() },
    val layoutPositions: Int = 0,
    val sectors: Int = 1,
    val info: LayoutInfo = LayoutInfo()
) {
    companion object

    val totalLayers = characterSets
        .indexOfLast { it.isNotEmpty() }
        .let { if (it == -1) 0 else it + 1 }

    override fun toString(): String {
        val sb = StringBuilder(info.name)
        if (totalLayers > 1) {
            sb.append(" (")
            sb.append(totalLayers)
            sb.append(" layers)")
        }
        return sb.toString()
    }
}

fun KeyboardData.addAllToActionMap(
    actionMapAddition: Map<MovementSequence, KeyboardAction>
): KeyboardData {
    return KeyboardData.actionMap.modify(this) {
        it + actionMapAddition
    }
}

fun KeyboardData.lowerCaseCharacters(layer: LayerLevel): Option<String> {
    return characterSets.elementAtOrNone(layer.ordinal - 1).flatMap {
        if (it.lowerCaseCharacters.isEmpty()) {
            none()
        } else {
            it.lowerCaseCharacters.some()
        }
    }
}

fun KeyboardData.upperCaseCharacters(layer: LayerLevel): Option<String> {
    return characterSets.elementAtOrNone(layer.ordinal - 1).flatMap {
        if (it.upperCaseCharacters.isEmpty()) {
            none()
        } else {
            it.upperCaseCharacters.some()
        }
    }
}

fun KeyboardData.setLowerCaseCharacters(
    lowerCaseCharacters: String,
    layer: LayerLevel
): KeyboardData {
    return KeyboardData.characterSets.index(
        Index.list(),
        layer.ordinal - 1
    ).lowerCaseCharacters.set(
        this,
        lowerCaseCharacters
    )
}

fun KeyboardData.setUpperCaseCharacters(
    upperCaseCharacters: String,
    layer: LayerLevel
): KeyboardData {
    return KeyboardData.characterSets.index(
        Index.list(),
        layer.ordinal - 1
    ).upperCaseCharacters.set(
        this,
        upperCaseCharacters
    )
}

fun KeyboardData.findLayer(movementSequence: MovementSequence): LayerLevel {
    return actionMap[movementSequence]?.layer.let {
        if (it == LayerLevel.HIDDEN) {
            LayerLevel.FIRST
        } else {
            it
        }
    } ?: LayerLevel.FIRST
}

fun KeyboardData.oppositeDirection(direction: Int): Int {
    return ((direction - 1) + sectors / 2) % sectors + 1
}

fun KeyboardData.computeSectorsAndLayouts(layers: Collection<Layer>): KeyboardData {
    val sectorsAndLayouts =
        layers.fold(sectors to layoutPositions) { acc, layer ->
            layer.sectors.fold(acc) { (maxSector, maxLayoutPositions), (sector, value) ->
                val layoutPositions = value.parts.values.maxByOrNull { it.size }?.size ?: 0
                maxSector.coerceAtLeast(sector) to maxLayoutPositions.coerceAtLeast(
                    layoutPositions
                )
            }
        }
    return copy(sectors = sectorsAndLayouts.first, layoutPositions = sectorsAndLayouts.second)
}

fun KeyboardData.characterSetSize(): Int {
    return sectors * 2 * layoutPositions
}
